Added option to send booleans as numbers rather than text.

RichardBH opened the pull request:
For example: rather than sending a coil as True/False send it as 1/0.

RichardBH committed:
Updated for input_register_32BE

Other spelling mistakes etc.

RichardBH committed:
Updated with input_register_32BE

And other clarifications.

RichardBH committed:
Fixed "Coil" that came back...

RichardBH committed:
Merge pull request #2 from mbs38/master

Pull down changes

RichardBH committed:
Added option to send bools as 0/1 rather than false/true

Added option to send bools as 0/1 rather than false/true, Added Holding Register 32.
Fixed 2 words to 1 long calculation.

mbs38 commented:
> For example: rather than sending a coil as True/False send it as 1/0.
 
I have already been thinking about this for a while. Thx for implementing it. 

However while we're at it, maybe we could do it in a more universal way and allow arbitrary values:
- add a command line arguments like --define-true and --define-false, store that to valTrue and valFalse (that obviously should default to True and False).
- where you are using 0 or 1 atm we could just use valTrue and valFalse instead (lines 277 and 279)

RichardBH commented:
Sure.

I have nearly finished the configparser rewrite and its a pretty big
change... I have maintained backward compatibility to arguments, I just
need to do more testing to ensure I haven't broken the arguments.

I have also rewritten  def poll(self): by adding more error checking,
(there is a feature/bug in pymodbus where it returns an exception object
rather than a result obect if there is an error. So this detects that and
gives more meaningful error messages.

I've made "with open(args.config,"r") as csvfile:" into a function with try
except blocks for better error handling.

And I've added an option to send the data as a JSON string with timestamp
rather than just topic/value.
   # Builds Payload and topíc in simple JSON parameters: DeviceID, Site,
Key, value, timestamp
   #{"DeviceId": <Device.name>, "Site':<Globaltopic>, "<ReferenceTopic>":
<Value>, "timestamp": 1548940870660}
   #{"DeviceId": "mez_master", "Site':<Globaltopic>, "C3_DOOR": 0,
"timestamp": 1548940870660}

I've also moved some global variables around.
I don't think I've broken anything and it should all work without any
changes to arguments. Just need to add the changes that Buskebam did with
the MQTT TLS into the config.ini file.

I'm glad he did it as it was on my todo list... :)

Is there a better way for the three of us to discuss future changes, as he
probably won't see this post.

I have all the above working in production, just need to test backward
compatibility.

BTW. With the config file, if the file modbus2mqtt.ini exists in the
current directory, it will ignore all arguments and read from the file.

I have attached the current versions to this email. (I haven't committed it
yet.)

Cheers,
Richard.





On Tue, Feb 5, 2019 at 1:47 PM Max B. <notifications@github.com> wrote:

> For example: rather than sending a coil as True/False send it as 1/0.
>
> I have already been thinking about this for a while. Thx for implementing
> it.
>
> However while we're at it, maybe we could do it in a more universal way
> and allow arbitrary values:
>
>    - add a command line arguments like --define-true and --define-false,
>    store that to valTrue and valFalse (that obviously should default to True
>    and False).
>    - where you are using 0 or 1 atm we could just use valTrue and
>    valFalse instead (lines 277 and 279)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mbs38/spicierModbus2mqtt/pull/7#issuecomment-460503137>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHui8WXXeBurGwHv9s5NN91ySEavmpg4ks5vKP7LgaJpZM4aidvT>
> .
>

# spicierModbus2mqtt - Modbus TCP/RTU to MQTT bridge (and vice versa)
# https://github.com/mbs38/spicierModbus2mqtt 
#
# Written in 2018 by Max Brueggemann <mail@maxbrueggemann.de>
#  
#
# Provided under the terms of the MIT license.

# Contains a bunch of code taken from:
# modbus2mqtt - Modbus master with MQTT publishing
# Written and (C) 2015 by Oliver Wagner <owagner@tellerulam.com>
# Provided under the terms of the MIT license.

# Main improvements over modbus2mqtt:
# - more abstraction when writing to coils/registers using mqtt. Writing is now
#   possible without having to know slave id, reference, function code etc.
# - specific coils/registers can be made read only
# - multiple slave devices on one bus are now supported
# - polling speed has been increased sgnificantly. With modbus RTU @ 38400 baud
#   more than 80 transactions per second have been achieved.
# - switched over to pymodbus which is in active development


# Requires:
# - Eclipse Paho for Python - http://www.eclipse.org/paho/clients/python/
# - pymodbus - https://github.com/riptideio/pymodbus

import argparse
import time
import socket
import paho.mqtt.client as mqtt
import serial
import io
import sys
import csv
import signal
import random
import ssl
import configparser
import os
import json

import addToHomeAssistant

from pymodbus.pdu import ModbusRequest
from pymodbus.client.sync import ModbusSerialClient as SerialModbusClient
from pymodbus.client.sync import ModbusTcpClient as TCPModbusClient
from pymodbus.transaction import ModbusRtuFramer

version="0.4.2"
parser = argparse.ArgumentParser(description='Bridge between ModBus and MQTT')

#Default globals
verbosity=1
addToHass=False




def signal_handler(signal, frame):
        print('Exiting ' + sys.argv[0])
        control.stopLoop()
signal.signal(signal.SIGINT, signal_handler)

class Control:
    def __init__(self):
        self.runLoop = True
    def stopLoop(self):
        self.runLoop = False


class Device:
    def __init__(self,name,slaveid):
        self.name=name
        self.occupiedTopics=[]
        self.writableReferences=[]
        self.slaveid=slaveid
        if verbosity>=2:
            print('Added new device \"'+self.name+'\"')


class Poller:
    def __init__(self,topic,rate,slaveid,functioncode,reference,size,dataType):
        self.topic=topic
        self.rate=float(rate)
        self.slaveid=int(slaveid)
        self.functioncode=int(functioncode)
        self.dataType=dataType
        self.reference=int(reference)
        self.size=int(size)
        self.next_due=time.clock_gettime(0)+self.rate*random.uniform(0,1)
        self.last = None
        self.readableReferences=[]
        self.device=None
        self.disabled=False
        self.failcounter=0

        for myDev in deviceList:
            if myDev.name == self.topic:
                self.device=myDev
                break
        if self.device == None:
            device = Device(self.topic,slaveid)
            deviceList.append(device)
            self.device=device
        if verbosity>=2:
            print("Added new Poller "+str(self.topic)+","+str(self.functioncode)+","+str(self.dataType)+","+str(self.reference)+","+str(self.size)+",")


    def failCount(self,failed):
        if not failed:
            self.failcounter=0
        else:
            if self.failcounter==3:
                self.disabled=True
                print("Poller "+self.topic+" with Slave-ID "+str(self.slaveid)+ " and functioncode "+str(self.functioncode)+" disabled due to the above error.")
            else:
                self.failcounter=self.failcounter+1


    def poll(self):
            result = None
            if master.is_socket_open()==True:
                try:
                    if self.functioncode == 1:
                        result = master.read_coils(self.reference, self.size, unit=self.slaveid)
                    elif self.functioncode == 2:
                        result = master.read_discrete_inputs(self.reference, self.size, unit=self.slaveid)
                    elif self.functioncode == 3:
                        result = master.read_holding_registers(self.reference, self.size, unit=self.slaveid)
                    elif self.functioncode == 4:
                        result = master.read_input_registers(self.reference, self.size, unit=self.slaveid)
                    #print("modbus result: "+str(result)+" Type: "+ str(type(result))+" FC: "+str(result.function_code))        
                    
                    if not result.isError():
                    # proceed
                        if result.function_code < 0x80:
                            if result.function_code <= 2: 
                                data = result.bits
                            elif 3 <= result.function_code < 0x80:    
                                data = result.registers
                            if verbosity>=4:
                                print("Read MODBUS, FC:"+str(self.functioncode)+", DataType:"+str(self.dataType)+", ref:"+str(self.reference)+", Qty:"+str(self.size)+", SI:"+str(self.slaveid))
                                print("Read MODBUS, DATA:"+str(data))
                            for ref in self.readableReferences:
                                ref.checkPublish(data,self.topic)
                        else:
                        # handle error or raise
                            if verbosity>=1:
                                print("Slave device "+str(self.slaveid)+" responded with error code: "+str(result.function_code))
                    else:
                    # handle error or raise
                        if verbosity>=1:
                            print("Slave device "+str(self.slaveid)+" responded with error code: "+str(result))
                except Exception as e:
                    failed = True
                    if verbosity>=1:
                        print("Error talking to slave device:"+str(self.slaveid)+", "+str(e))
                if args.autoremove:
                    self.failCount(failed)
            else:
                if master.connect():
                    if verbosity >= 1:
                        print("MODBUS connected successfully")
                else:
                    if verbosity >= 1:
                        print("MODBUS connection error, trying again...")

    def checkPoll(self):
        if time.clock_gettime(0) >= self.next_due and not self.disabled:
            self.poll()
            self.next_due=time.clock_gettime(0)+self.rate


    def addReference(self,myRef):
        #check reference configuration and maybe add to this poller or to the list of writable things
        if myRef.topic not in self.device.occupiedTopics:
            self.device.occupiedTopics.append(myRef.topic)
            if "r" in myRef.rw or "w" in myRef.rw:
                myRef.device=self.device
                if "r" in myRef.rw:
                    if myRef.checkSanity(self.reference,self.size):
                        self.readableReferences.append(myRef)
                        if "w" not in myRef.rw:
                            referenceList.append(myRef)
                            if verbosity >= 2:
                                print('Added new reference \"' + myRef.topic + '\"')

                    else:
                        print("Reference \""+str(myRef.reference)+"\" with topic "+myRef.topic+" is not in range ("+str(self.reference)+" to "+str(int(self.reference+self.size-1))+") of poller \""+self.topic+"\", therefore ignoring it for polling.")
                if "w" in myRef.rw:
                    if self.functioncode == 3: #holding registers
                        myRef.writefunctioncode=6 #preset single register
                    if self.functioncode == 1: #coils
                        myRef.writefunctioncode=5 #force single coil
                    if self.functioncode == 2: #read input status, not writable
                        print("Reference \""+str(myRef.reference)+"\" with topic "+myRef.topic+" in poller \""+self.topic+"\" is not writable (discrete input)")
                    if self.functioncode == 4: #read input register, not writable
                        print("Reference \""+str(myRef.reference)+"\" with topic "+myRef.topic+" in poller \""+self.topic+"\" is not writable (input register)")
                    if myRef.writefunctioncode is not None:
                       self.device.writableReferences.append(myRef)
                       referenceList.append(myRef)
            else:
                print("Reference \""+str(myRef.reference)+"\" with topic "+myRef.topic+" in poller \""+self.topic+"\" is neither read nor writable, therefore ignoring it.")
        else:
            print("Reference topic ("+str(myRef.topic)+") is already occupied for poller \""+self.topic+"\", therefore ignoring it.")

class Reference:
    def __init__(self,topic,reference,format,rw,poller):
        self.topic=topic
        self.reference=int(reference)
        self.format=format.split(":",2)
        self.lastval=None
        self.rw=rw
        self.relativeReference=None
        self.writefunctioncode=None
        self.device=None
        self.poller=poller
        self.value = 0

    def checkSanity(self,reference,size):
        if self.reference in range(reference,size+reference):
            self.relativeReference=self.reference-reference
            return True

    def checkPublish(self,result,topic):
        # Only publish messages after the initial connection has been made. If it became disconnected then the offline buffer will store messages,
        # but only after the intial connection was made.
        if mqc.initial_connection_made == True:
            if self.poller.dataType == "int32":
                val = result[self.relativeReference]*65536 + result[self.relativeReference+1]
            elif self.poller.dataType == "int32":
                val = result[self.relativeReference]
            elif self.poller.dataType == "bool":
                if args.bool_number == True: #We need to change True = 1 and False = 0
                    if result[self.relativeReference] == True: 
                        val = 1
                    elif result[self.relativeReference] == False:
                        val = 0
                else:
                    val = result[self.relativeReference]
            else:
                val = result[self.relativeReference]
                
            if self.lastval != val or integrity_send == True:
                self.lastval = val

                if args.payload_style.casefold() == 'JSON'.casefold():
                    try:
                        # Builds Payload and topíc in simple JSON parameters: DeviceID, Key, value, timestanmp
                        #{"DeviceId": <Device.name>, "Site':<Globaltopic>, "<ReferenceTopic>": <Value>, "timestamp": 1548940870660}                        
                        #{"DeviceId": "mez_master", "Site':<Globaltopic>, "C3_DOOR": 0, "timestamp": 1548940870660}                        
                                              
                        payload = json.dumps({"DeviceId":self.device.name, "Site":globaltopic,self.topic: val,"timestamp":timestamp_unix_ms})
                        publish_result = mqc.publish(globaltopic, payload,qos=1,retain=True)
                        if verbosity>=4:
                            print("MQTT T:" + str(globaltopic)+",PL: " + str(payload)+" Result:"+str(publish_result.rc))
                    except Exception as e:
                        if verbosity>=1:
                            print("Error publishing JSON MQTT {}".format(e))

                else:
                    try:
                        publish_result = mqc.publish(globaltopic+self.device.name+"/state/"+self.topic,val,qos=1,retain=True)
                        if verbosity>=4:
                            print("published MQTT topic: " + str(globaltopic+self.device.name+"/state/"+self.topic)+" value: " + str(val)+" RC:"+str(publish_result.rc))
                    except Exception as e:
                        if verbosity>=1:
                            print("Error publishing MQTT topic: " + str(self.device.name+"/state/"+self.topic)+" value: " + str(val)+","+str(e) )

def handleArguments():
#Check if the modbus2mqtt.ini file exists, if it does then read values from there and ignore arguments.
#If not then parse the arguments.
    try:    
        print("parsing arguments.")
        #setup and parse arguments
        parser.add_argument('--mqtt-host', default='localhost', help='MQTT server address. Defaults to "localhost"')
        parser.add_argument('--mqtt-port', default='1883', type=int, help='MQTT server port. Defaults to 1883')
        parser.add_argument('--mqtt-topic', default='modbus/', help='Topic prefix to be used for subscribing/publishing. Defaults to "modbus/"')
        parser.add_argument('--mqtt-user', default=None, help='Username for authentication (optional)')
        parser.add_argument('--mqtt-pass', default="", help='Password for authentication (optional)')
        parser.add_argument('--rtu',help='pyserial URL (or port name) for RTU serial port')
        parser.add_argument('--rtu-baud', default='19200', type=int, help='Baud rate for serial port. Defaults to 19200')
        parser.add_argument('--rtu-parity', default='even', choices=['even','odd','none'], help='Parity for serial port. Defaults to even')
        parser.add_argument('--tcp', help='Act as a Modbus TCP master, connecting to host TCP')
        parser.add_argument('--tcp-port', default='502', type=int, help='Port for MODBUS TCP. Defaults to 502')
        parser.add_argument('--set-modbus-timeout',default='1',type=float, help='Response time-out for MODBUS devices')
        parser.add_argument('--config', required=True, help='Configuration file. Required!')
        parser.add_argument('--verbosity', default='3', type=int, help='Verbose level, 0=silent, 1=errors only, 2=connections, 3=mb writes, 4=all')
        parser.add_argument('--autoremove',action='store_true',help='Automatically remove poller if modbus communication has failed three times.')
        parser.add_argument('--add-to-homeassistant',action='store_true',help='Add devices to Home Assistant using Home Assistant\'s MQTT-Discovery')
        parser.add_argument('--payload-style', default='std', choices=['std','json'], help='MQTT payload type. std is just the number. json is JSON. Defaults to std (optional)')
        parser.add_argument('--set-loop-break',default='0.01',type=float, help='Set pause in main polling loop. Defaults to 10ms.')
        parser.add_argument('--bool-number',default=False, type=bool, help='Send bools, e.g. coil as a number 0 or 1' )
        args=parser.parse_args()
        verbosity=args.verbosity
        addToHass=args.add_to_homeassistant
        #Tihs is here for backwards compatibility, where rtu previously had prioty over tcp mode.    
        if args.rtu:
            args.modbus_mode = 'rtu'
        else:
            args.modbus_mode = 'tcp'
    except Exception as e:
        if verbosity >= 1:
            print("Error handling arguments: " + str(e))
        exit(1)
            

class argsFromINI:
    def __init__(self):
        self.mqtt_host      = 'localhost'
        self.mqtt_port      = int(1883)
        self.mqtt_topic     ='modbus/'
        self.mqtt_user      =None
        self.mqtt_pass      =""
        self.rtu            = '/dev/ttyS0'
        self.rtu_baud       =int(19200)
        self.rtu_parity     ='even'
        self.tcp            = 'localhost'
        self.tcp_port       = int(502)
        self.set_modbus_timeout=float(1.0)
        self.config         ='config.csv'
        self.verbosity      = int(3)
        self.autoremove     = False
        self.add_to_homeassistant = False
        self.payload_style   ='std'
        self.set_loop_break = float(0.01)
        self.bool_number    = False
        self.modbus_mode    = 'rtu'

def readINIConfigFile():
#Returns 0 if file exists and no errors.
#returns 1 if file does not exist.
#returns 2 if there were errors reading the INI
    try:
        config_INI_filename = 'modbus2mqtt.ini'
        config_INI_path     = sys.path[0]+'/'+config_INI_filename
        print("Opening INI fie: " + config_INI_path)
        inifile_exists = os.path.isfile(config_INI_path)
        if inifile_exists:
            #configparser Read in whole INI file.
            cfg = configparser.ConfigParser()
            result = cfg.read(config_INI_path)
            if result:
                #setup all arguments.
                args.modbus_mode = cfg.get('MODBUS_INPUT','modbus_mode')
                print("modbus_mode:" +args.modbus_mode)
                args.mqtt_host  = cfg.get('MQTT_OUTPUT','mqtt_host')
                args.mqtt_port  = cfg.getint('MQTT_OUTPUT','mqtt_port')
                args.mqtt_topic = cfg.get('MQTT_OUTPUT','mqtt_topic')
                args.mqtt_user  = cfg.get('MQTT_OUTPUT','mqtt_user')
                args.mqtt_pass  = cfg.get('MQTT_OUTPUT','mqtt_pass')
                args.payload_style   = cfg.get('MQTT_OUTPUT','payload_style')
                args.bool_number = cfg.getboolean('MQTT_OUTPUT','bool_number')

                args.rtu        = cfg.get('MODBUS_INPUT','rtu')
                args.rtu_baud   = cfg.getint('MODBUS_INPUT','rtu_baud')
                args.rtu_parity = cfg.get('MODBUS_INPUT','rtu_parity')
                args.tcp        = cfg.get('MODBUS_INPUT','tcp')
                print("tcp:" +args.tcp)
                args.tcp_port   = cfg.getint('MODBUS_INPUT','tcp_port')
                args.set_modbus_timeout     = cfg.getfloat('MODBUS_INPUT','set_modbus_timeout')
                args.config                 = cfg.get('MODBUS_INPUT','config')

                args.autoremove             = cfg.getboolean('MODBUS_INPUT','autoremove')
                args.verbosity              = cfg.getint('DEBUG','verbosity')
                args.add_to_homeassistant   = cfg.getboolean('HOME_ASSISTANT','add_to_homeassistant')
                args.set_loop_break         = cfg.getfloat('TIMING','set_loop_break')
                print("args: " + str(args))
                return 0
        else:
            return 1
    except Exception as e:
        if verbosity >= 1:
            print("Error opening INI file: " + str(e))
        return 2

                        

def readModbusConfigFile():
    # Now let's read the config file
    try: 
        print("modbus config:" +sys.path[0]+'/'+args.config)
        with open(sys.path[0]+'/'+args.config,"r") as csvfile:
            csvfile.seek(0)
            reader=csv.DictReader(csvfile)
            currentPoller=None
            for row in reader:
                if row["type"]=="poller" or row["type"]=="poll":
                    if row["col5"] == "holding_register":
                        functioncode = 3
                        dataType="int16"
                    if row["col5"] == "holding_register_32BE":
                        functioncode = 3
                        dataType="int32"
                    if row["col5"] == "coil":
                        functioncode = 1
                        dataType="bool"
                    if row["col5"] == "input_register":
                        functioncode = 4
                        dataType="int16"
                    if row["col5"] == "input_status":
                        functioncode = 2
                        dataType="bool"
                    if row["col5"] == "input_register_32BE":
                        functioncode = 4
                        dataType="int32"
                    rate = float(row["col6"])
                    slaveid = int(row["col2"])
                    reference = int(row["col3"])
                    size = int(row["col4"])
                    currentPoller = Poller(row["topic"],rate,slaveid,functioncode,reference,size,dataType)
                    pollers.append(currentPoller)
                    continue
                elif row["type"]=="reference" or row["type"]=="ref":
                    reference = int(row["col2"])
                    currentPoller.addReference(Reference(row["topic"],reference,"",row["col3"],currentPoller))
    except Exception as e:
        if verbosity >= 1:
            print("Error opening Modbus CSV file, this file must exist: " + str(e))
        exit(1)

def messagehandler(mqc,userdata,msg):
    if True:
        (prefix,device,function,reference) = msg.topic.split("/")
        if function != 'set':
            return
        myRef = None
        myDevice = None
        for iterDevice in deviceList:
            if iterDevice.name == device:
                myDevice = iterDevice
        if myDevice == None: # no such device
            return
        for iterRef in myDevice.writableReferences:
            if iterRef.topic == reference:
                myRef=iterRef
        if myRef == None: # no such reference
            return    
        payload = str(msg.payload.decode("utf-8"))
        if myRef.writefunctioncode == 5:
            value = None
            if payload == 'True' or payload == 'true' or payload == '1' or payload == 'TRUE':
                value = True
            if payload == 'False' or payload == 'false' or payload == '0' or payload == 'FALSE':
                value = False
            if value != None:
                    result = master.write_coil(int(myRef.reference),value,unit=int(myRef.device.slaveid))
                    try:
                        if result.function_code < 0x80:
                            if verbosity>=3:
                                print("Writing to device "+str(myDevice.name)+", Slave-ID="+str(myDevice.slaveid)+" at Reference="+str(myRef.reference)+" using function code "+str(myRef.writefunctioncode)+" successful.")
                        else:
                            if verbosity>=1:
                                print("Writing to device "+str(myDevice.name)+", Slave-ID="+str(myDevice.slaveid)+" at Reference="+str(myRef.reference)+" using function code "+str(myRef.writefunctioncode)+" FAILED! (Devices responded with errorcode. Maybe bad configuration?)")
            
                    except NameError:
                        if verbosity>=1:
                            print("Error writing to slave device "+str(myDevice.slaveid)+" (maybe CRC error or timeout)")
            else:
                if verbosity >= 1:
                    print("Writing to device "+str(myDevice.name)+", Slave-ID="+str(myDevice.slaveid)+" at Reference="+str(myRef.reference)+" using function code "+str(myRef.writefunctioncode)+" not possible. Given value is not \"True\" or \"False\".")


        if myRef.writefunctioncode == 6:
            try:
                value=int(payload)
                if value > 65535 or value < 0:
                    value = None
            except Exception as e:
                value=None
            
                    
            if value is not None:
                result = master.write_registers(int(myRef.reference),value,unit=myRef.device.slaveid)
                try:
                    if result.function_code < 0x80:
                        if verbosity>=3:
                            print("Writing to device "+str(myDevice.name)+", Slave-ID="+str(myDevice.slaveid)+" at Reference="+str(myRef.reference)+" using function code "+str(myRef.writefunctioncode)+" successful.")
                    else:
                        if verbosity>=1:
                            print("Writing to device "+str(myDevice.name)+", Slave-ID="+str(myDevice.slaveid)+" at Reference="+str(myRef.reference)+" using function code "+str(myRef.writefunctioncode)+" FAILED! (Devices responded with errorcode. Maybe bad configuration?)")
                except NameError:
                    if verbosity >= 1:
                        print("Error writing to slave device "+str(myDevice.slaveid)+" (maybe CRC error or timeout)")
            else:
                if verbosity >= 1:
                    print("Writing to device "+str(myDevice.name)+", Slave-ID="+str(myDevice.slaveid)+" at Reference="+str(myRef.reference)+" using function code "+str(myRef.writefunctioncode)+" not possible. Given value is not an integer between 0 and 65535.")
        
def connecthandler(mqc,userdata,flags,rc):
    if rc == 0:
        mqc.initial_connection_made = True
        if verbosity>=2:
            print("MQTT Broker connected succesfully: " + args.mqtt_host + ":" + str(mqtt_port))
        mqc.subscribe(globaltopic + "+/set/+")
        if verbosity>=2:
            print("Subscribed to MQTT topic: "+globaltopic + "+/set/+")
        mqc.publish(globaltopic + "connected", "True", qos=1, retain=True)
    elif rc == 1:
        if verbosity>=1:
            print("MQTT Connection refused – incorrect protocol version")
    elif rc == 2:
        if verbosity>=1:
            print("MQTT Connection refused – invalid client identifier")
    elif rc == 3:
        if verbosity>=1:
            print("MQTT Connection refused – server unavailable")
    elif rc == 4:
        if verbosity>=1:
            print("MQTT Connection refused – bad username or password")
    elif rc == 5:
        if verbosity>=1:
            print("MQTT Connection refused – not authorised")

def disconnecthandler(mqc,userdata,rc):
    if verbosity >= 2:
        print("MQTT Disconnected, RC:"+str(rc))

def loghandler(mgc, userdata, level, buf):
    if verbosity >= 4:
        print("MQTT LOG:" + buf)



if result == 1: #file doesnt exist so handle arguments
    print("handling arguments")
    handleArguments()
elif result == 2:
    if verbosity >=1:
        print("Error reading INI file.")
    exit(1)
print("modbus_mode:" +args.modbus_mode)

verbosity                   = args.verbosity
addToHass                   = args.add_to_homeassistant
 
integrity_send = False
integrity_send_rate = 600
integrity_send_next_due = 0

control = Control()
print("modbus_mode:" +args.modbus_mode)

globaltopic=args.mqtt_topic

if not globaltopic.endswith("/"):
    globaltopic+="/"

if verbosity>=0:
    print('Starting spiciermodbus2mqtt V%s with topic prefix \"%s\"' %(version, globaltopic))
    
master=None

deviceList=[]
referenceList=[]

pollers=[]

# type, topic, slaveid,   ref, size, functioncode, rate
# type, topic, reference, rw,      ,             ,

timestamp_unix_ms = int(time.time() * 1000)
print("b4 d modbus_mode:" +args.modbus_mode)

readModbusConfigFile()
print("after md modbus_mode:" +args.modbus_mode)

if True:
#Setup MODBUS Master
    print("modbus_mode if: "+args.modbus_mode)
    if args.modbus_mode == 'rtu':
    
        if args.rtu_parity == "none":
            parity = "N"
        if args.rtu_parity == "odd":
            parity = "O"
        if args.rtu_parity == "even":
            parity = "E"

        print("About to Connect RTU serial")
        master = SerialModbusClient(method="rtu", port=args.rtu, stopbits = 1, bytesize = 8, parity = parity, baudrate = int(args.rtu_baud), timeout=args.set_modbus_timeout)

    elif args.tcp:
        print("About to connect TCP modbus")
        master = TCPModbusClient(args.tcp, args.tcp_port,client_id="foo123", clean_session=False, timeout=args.set_modbus_timeout)
        print("MODBUS connected: "+str(master))

    else:
        print("You must specify a modbus access method, either --rtu or --tcp")
        sys.exit(1)

#Setup MQTT Broker

    mqtt_port = args.mqtt_port

    if mqtt_port is None:
        if args.mqtt_use_tls:
            mqtt_port = 8883
        else:
            mqtt_port = 1883

    clientid=globaltopic + "-" + str(time.time())
    mqc=mqtt.Client(client_id=clientid)
    mqc.on_connect=connecthandler
    mqc.on_message=messagehandler
    mqc.on_disconnect=disconnecthandler
    mqc.on_log= loghandler
    mqc.will_set(globaltopic+"connected","True",qos=2,retain=True)
    mqc.initial_connection_attempted = False
    mqc.initial_connection_made = False
    if args.mqtt_user or args.mqtt_pass:
        mqc.username_pw_set(args.mqtt_user, args.mqtt_pass)

    if args.mqtt_use_tls:

        if args.mqtt_tls_version == "tlsv1.2":
            tls_version = ssl.PROTOCOL_TLSv1_2
        elif args.mqtt_tls_version == "tlsv1.1":
            tls_version = ssl.PROTOCOL_TLSv1_1
        elif args.mqtt_tls_version == "tlsv1":
            tls_version = ssl.PROTOCOL_TLSv1
        elif args.mqtt_tls_version is None:
            tls_version = None
        else:
            if verbosity >= 2:
                print("Unknown TLS version - ignoring")
            tls_version = None


        if args.mqtt_insecure:
            cert_regs = ssl.CERT_NONE
        else:
            cert_regs = ssl.CERT_REQUIRED

        mqc.tls_set(ca_certs=args.mqtt_cacerts, certfile= None, keyfile=None, cert_reqs=cert_regs, tls_version=tls_version)

        if args.mqtt_insecure:
            mqc.tls_insecure_set(True)



#Setup HomeAssistant
    if(addToHass):
        adder=addToHomeAssistant.HassConnector(mqc,globaltopic,verbosity)
        adder.addAll(referenceList)

#Main Loop
modbus_connected = False
while control.runLoop:
    timestamp_unix_ms = int(time.time() * 1000)
    if not modbus_connected:
        print("Connecting to MODBUS...")
        modbus_connected = master.connect()
        if modbus_connected:
            if verbosity >= 2:
                print("MODBUS connected successfully")
        else:
            if verbosity >= 1:
                print("MODBUS connection error, trying again...")

    if not mqc.initial_connection_attempted:
       try:
            print("Connecting to MQTT Broker: " + args.mqtt_host + ":" + str(mqtt_port) + "...")
            mqc.connect(args.mqtt_host, mqtt_port, 60)
            mqc.initial_connection_attempted = True #Once we have connected the mqc loop will take care of reconnections.
            mqc.loop_start()
            if verbosity >= 1:
                print("MQTT Loop started")
        except Exception as e:
            if verbosity>=1:
              print("Socket Error connecting to MQTT broker: " + args.mqtt_host + ":" + str(mqtt_port) + ", check LAN/Internet connection, trying again...")

    if mqc.initial_connection_made: #Don't start polling unless the initial connection to MQTT has been made, no offline MQTT storage will be available until then.
        if modbus_connected:
            try:
                for p in pollers:
                    p.checkPoll()
            except Exception as e:
                if verbosity>=1:
                    print("Exception Error when polling or publishing, trying again...")

    try:
        time.sleep(args.set_loop_break)
        integrity_send = False
        if time.clock_gettime(0) >= integrity_send_next_due:
            integrity_send = True
            integrity_send_next_due=time.clock_gettime(0)+integrity_send_rate
    except Exception as e:
        if verbosity>=1:
            print("Error when determining next integrity due time" + str(e))
master.close()
#adder.removeAll(referenceList)
sys.exit(1)

mbs38 commented:
Well there is still a lot to do :-)

> #Check if the modbus2mqtt.ini file exists, if it does then read values from there and ignore arguments.

about this: I would like to see it the other way round because this can easily break running installations. Also for testing it is neat to override a setting by calling the script with an argument.

Best,
Max

mbs38 commented:
Sorry wrong button... xD

RichardBH commented:
Ok,

How about I add the ini file as an argument like the config file.
So at minimum to run it you need to still supply the --config csv filename.

I'll add a --ini argument.

Where the arguments will take preference over the ini file contents.
My only unknown is that how do you know if an argument was supplied or if
it was a default?

Cheers,
Richard.

On Tue, Feb 5, 2019 at 3:44 PM Max B. <notifications@github.com> wrote:

> Well there is still a lot to do :-)
>
> #Check if the modbus2mqtt.ini file exists, if it does then read values
> from there and ignore arguments.
> about this: I would like to see it the other way round because this can
> easily break running installations. Also for testing it is neat to override
> a setting by calling the script with an argument.
>
> Best,
> Max
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mbs38/spicierModbus2mqtt/pull/7#issuecomment-460520370>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHui8d1WeLjdVwJtlXL_8p3ElRwEQTdoks5vKRoogaJpZM4aidvT>
> .
>

mbs38 commented:
Well we can do it by hand: defaulting to None and checking afterwards

parser.add_argument('--rtu-baud', default=None, type=int, help='Baud rate for serial port. Defaults to 19200')
[...]

if args.rtu_baud not None:
    rtu_baud=19200
else: #use ini file content instead
    [...]

and then use rtu_baud later on..
Of course this might become a little messy.

mbs38 commented:
I have added some more features you might want to take a look at. Sorry for breaking "input_register_32BE", but I think that the intepretation of data should be defined in the reference object.
Max

bikeymouse commented:
> I have added some more features you might want to take a look at. Sorry for breaking "input_register_32BE", but I think that the intepretation of data should be defined in the reference object.
> Max

Hi,

I'm not sure if my issue is related to this, but I'm having trouble to get proper readings from a SDM630 energy meter.

It uses 2 registers for each reading, but these must be combined to form a floating point (IEEE 754) number, with the most significant register first. Unfortunately I only get strange readings back.

Could anyone help to indicate if this can be read with spicierModbus2mqtt and what the config should look like? Is this a situation where I would have to use the "input_register_32BE" format?

As and example, I'm using this in the config to read the voltage for 3 phases:
poll,modbus,1,0,6,input_register,3.0
ref,voltagephase1,0,r,int32BE
ref,voltagephase2,2,r,int32BE
ref,voltagephase3,4,r,int32BE

The description of the registers and Modbus format for this meter is described [here](http://www.eastrongroup.com/data/uploads/Eastron_SDM630MV_CT_protocol_V1_0_.pdf).

Thanks in advance for any tips!

mbs38 commented:
Hi bikeymouse,
> I'm not sure if my issue is related to this

It's not.

But you are actually hitting an issue here: Currently signed integers like int32BE are not supported (it's proposed :P) but even if they were, it would not work because your meter is using floats. In addition to that I think the register order is the other way round.

This is from the datasheet:
> In normal mode, the two registers that make up a floating point number are sent most significant register first.

That would be int32LE then.. but again this won't work because it's using floats..

Floats have not been implemented yet. Feel free to do so. Other than that your configuration looks fine to me.


Best,
Max

bikeymouse commented:
Hi Max,

Thanks for your reply. That explains why I didn't get any meaningful data ;-)
I'm not proficient at coding, so unfortunately I can't fix this myself.

Thanks anyway, I'll be watching this, so perhaps floats are implemented by somebody in the future ;-)

mbs38 commented:
> Thanks anyway, I'll be watching this, so perhaps floats are implemented by somebody in the future ;-)

Reading 32bit floats is now implemented (branch testing). I have no device to test this though. Can you try it please? :-)

Please try data type float32LE. If you get garbage try float32BE. Although I'm pretty sure your device needs float32LE.

Best,
Max

bikeymouse commented:
Hi,
Thanks, for this update! I have tried it and it seems to work.

The only thing is that I struggle to get data from registers when not using "0" as the base address.

For example, this works fine if I want to retrieve the register starting at base "0" and offset "4":
```
poll,test,1,0,6,input_register,2.0
ref,voltagephase3,4,r,float32LE
```
This works great I get good Voltage readings from my SDM630 meter.

However with the following setting in the config file, which would read the same register, but by starting at the offset of 4 and the index 0, gives no results on the mqqt-topic:
```
poll,test,1,4,2,input_register,2.0
ref,voltagephase3,0,r,float32LE
```
The log however does show that data values are being read from the modbus:

```
2020-01-01T21:04:39.556877644Z Read MODBUS, FC:4, DataType:int16, ref:4, Qty:2, SI:1
2020-01-01T21:04:39.556958791Z Read MODBUS, DATA:[17246, 4208]
2020-01-01T21:04:39.557039260Z Read MODBUS, FC:4, DataType:int16, ref:4, Qty:2, SI:1
2020-01-01T21:04:39.557123115Z Read MODBUS, DATA:[17246, 31987]
2020-01-01T21:04:39.557203011Z Read MODBUS, FC:4, DataType:int16, ref:4, Qty:2, SI:1
2020-01-01T21:04:39.557287074Z Read MODBUS, DATA:[17246, 20635]
2020-01-01T21:04:39.557368013Z Read MODBUS, FC:4, DataType:int16, ref:4, Qty:2, SI:1
2020-01-01T21:04:39.557471086Z Read MODBUS, DATA:[17246, 22890]
```
So, did I do something wrong, or is this a bug?

mbs38 commented:
> This works great I get good Voltage readings from my SDM630 meter.

That is good to hear.

> The only thing is that I struggle to get data from registers when not using "0" as the base address.

We use absolute addresses for the references:

```
poll,heating,32,52,4,holding_register,1
ref,temperature_post,53,r
ref,temperature_setpoint,54,rw
ref,temperature_outdoors,55,r
```

This is from the example.csv file. Something I should have included in the README too. Also now that I have read the README and example.csv files I think I really should do some cleanup xD

Best,
Max

bikeymouse commented:
Thanks for your reply, I have got it working now, great!
Thanks again for all your effort!

mbs38 commented:
You're welcome.
